server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -202,7 +203,8 @@ public DesiredBalance compute(
                         && routingAllocation.deciders()
                             .canAllocate(shardRouting, routingNode, routingAllocation)
                             .type() != Decision.Type.NO) {
-                        final var shardToInitialize = unassignedPrimaryIterator.initialize(nodeId, null, 0L, changes);
+                        final long expectedShardSize = getExpectedShardSize(shardRouting, 0L, routingAllocation);
+                        final var shardToInitialize = unassignedPrimaryIterator.initialize(nodeId, null, expectedShardSize, changes);
```

Comment:
This is also going to use additional disk space when initializing shards after SHRINK/SPLIT/CLONE, however this should not be the case according to this comment: https://github.com/elastic/elasticsearch/blob/cb6a570c49387ffa75098cf02f3df75ceaa61c3e/server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java#L131-L137

Reply:
I'll wait for this before approving.

Reply:
This is going to depend on the `shouldReserveSpaceForInitializingShard` added in https://github.com/elastic/elasticsearch/pull/102879